beets/metadata_plugins.py

```diff
@@ -0,0 +1,365 @@
+        dist.update(plugin.album_distance(items, album_info, mapping))
+    return dist
+
+
+def _get_distance(
+    config: ConfigView, data_source: str, info: AlbumInfo | TrackInfo
+) -> Distance:
+    """Returns the ``data_source`` weight and the maximum source weight
+    for albums or individual tracks.
+    """
+    from beets.autotag.hooks import Distance
+
+    dist = Distance()
+    if info.data_source == data_source:
+        dist.add("source", config["source_weight"].as_number())
+    return dist
+
+
+class MetadataSourcePlugin(BeetsPlugin, metaclass=abc.ABCMeta):
+    """A plugin that provides metadata from a specific source.
+
+    This base class implements a contract for plugins that provide metadata
+    from a specific source. The plugin must implement the methods to search for albums
+    and tracks, and to retrieve album and track information by ID.
+    """
+
+    data_source: str
+
+    def __init__(self, data_source: str, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.data_source = data_source or self.__class__.__name__
+        self.config.add({"source_weight": 0.5})
+
+    # --------------------------------- id lookup -------------------------------- #
+
+    def albums_for_ids(self, ids: Sequence[str]) -> Iterator[AlbumInfo | None]:
+        """Batch lookup of album metadata for a list of album IDs.
+
+        Given a list of album identifiers, yields corresponding AlbumInfo
+        objects. Missing albums result in None values in the output iterator.
+        Plugins may implement this for optimized batched lookups instead of
+        single calls to album_for_id.
+        """
+
+        return iter(self.album_for_id(id) for id in ids)
+
+    @abc.abstractmethod
+    def album_for_id(self, album_id: str) -> AlbumInfo | None:
+        """Return :py:class:`AlbumInfo` object or None if no matching release was
+        found."""
+        raise NotImplementedError
+
+    def tracks_for_ids(self, ids: Sequence[str]) -> Iterator[TrackInfo | None]:
+        """Batch lookup of track metadata for a list of track IDs.
+
+        Given a list of track identifiers, yields corresponding TrackInfo objects.
+        Missing tracks result in None values in the output iterator. Plugins may
+        implement this for optimized batched lookups instead of single calls to
+        track_for_id.
+        """
+
+        return iter(self.track_for_id(id) for id in ids)
+
+    @abc.abstractmethod
+    def track_for_id(self, track_id: str) -> TrackInfo | None:
+        """Return a :py:class:`AlbumInfo` object or None if no matching release was
+        found.
+        """
+        raise NotImplementedError
+
+    # ---------------------------------- search ---------------------------------- #
+
+    @abc.abstractmethod
+    def candidates(
+        self,
+        items: Sequence[Item],
+        artist: str,
+        album: str,
+        va_likely: bool,
+        extra_tags: dict[str, Any] | None = None,
```

Comment:
Note I removed `extra_tags` from all signatures in this PR: https://github.com/beetbox/beets/pull/5789/commits/d47558b2465160801129c3cf13d1e3a12044cdce as it's unused

Reply:
Interesting, I already thought they are mostly unused. Want me to remove it here too or do we want to prioritize #5789 first?

Reply:
As we merged the PR #5789 into master, I now removed all extra_tags in this PR too.